plugins/inventory-backend/README.md

```diff
@@ -0,0 +1,6 @@
+# Title
+Welcome to the inventory-backend plugin!
```

Comment:
The folder structure was a bit unexpected for me. Do we want to say that "plugins" are not just frontend? I'm not saying yes or no, but I do want this to be a conscious decision. It is important how we talk about different concepts in the repo.

Reply:
When I was looking though this for inspiration on how to do the scaffolder I though it would be pretty good to have a `backend` folder inside a plugin that you could require in the `backend` package. But because how currently `cjs` is setup, you can't have non webpackable modules in plugins if they're required in the client.

Reply:
yeah common node_modules space is tricky, and especially tricky if you want to try to combine in the same package as a plugin. Pretty sure the frontend and backend parts of a plugin need to be published separately with separate package.json, but there are many ways to achieve that.

Couple of options I can think of:

- separate folder:
  - `plugins/home-page`
  - `plugins/inventory`
  - `plugins/lighthouse`
  - `plugins-backend/inventory`
  - `plugins-backend/lighthouse`
- internal:
  - `plugins/home-page`
  - `plugins/inventory`
  - `plugins/inventory/backend`
  - `plugins/lighthouse`
  - `plugins/lighthouse/backend`
- separate workspace:
  - `frontend/plugins/home-page`
  - `frontend/plugins/inventory`
  - `frontend/plugins/lighthouse`
  - `backend/plugins/inventory`
  - `backend/plugins/lighthouse`

I think we should consider standalone plugin development outside of this repo early on as well. What would it look like for someone developing a single plugin in a separate repo, would they end up having to set up a monorepo as well? Tbh that might not be the end of the world as long as we have good tooling around it, but otherwise that concern would make us lean towards something where the backend is more integrated into the frontend package.

Reply:
Conceptually I thin the "internal" model is easier to wrap your head around.

Reply:
@stefanalund yep I agree, but it's the most technically challenging one. I think it's worth going down that route and see if we can figure it out though.

Reply:
Knowing nothing about the packaging setup in npm, but can't we nest it like so: `plugins/lighthouse/api` and `plugins/lighthouse/frontend`? Separate packages but we'll still have them colocated in the source tree?

That'd also extend to other plugin types, say if we add `plugins/lighthouse/events` for hooking into the event stream or `plugins/lighthouse/export` if we want to do a data dump.

Reply:
Yeah that's another option, but imo it's functionally similar to the separate folder or separate workspace structures. One thing you could do is wrap each of those in some parent tooling that lives inside e.g. `plugins/lighthouse`, stuff like "make a production build for both the frontend and backend", or "start up a development server for the frontend and backend and set up a local proxy to the backend". Tbh I think I'd rather promote the frontend part of the plugin to being the "main" part and have the backend in a dir.

Regarding npm packaging you couldn't have `plugins/lighthouse` as the published package, as that would leave us with the common deps between frontend and backend problem.

Reply:
I have a hard time seeing how we could publish just one package with both backend and frontend implementations in it, without hurting bundle size etc for the frontend. But I'm no expert at all.

To me, two packages right next to each other felt right as a first stab. The "separate workspaces" solution doesn't seem to add much as far as I can tell?

@benjdlambert do you have experience with this kind of mixing? It would be great to have it all in one package, if it were possible, of course. Is there something we can do to get to that state?

Reply:
I don't think you could publish a single package, but you might be able to publish 2 from 1 root dir.

Totally missed the second part of @mbruggmann comment though, I do like the idea of leaving the possibility for more types of deliverables associated with a plugin

Reply:
Yep, that last part makes sense. Would you like me to split this up in `inventory/frontend` and `inventory/backend`?

Does that affect the lerna setup? Its paths setup would need to change right?

Reply:
Tbh I'd rather merge the approach in this PR, since it's pretty similar anyway. Then we don't have to modify the lerna config and it becomes easier for more ppl to experiment

Reply:
@freben, So - without knowing too much about the background of why things are the way they are, I think the thing that really blocks everything is that the packages are built separately into `.cjs` files which is the entire bundle of the plugin or package. This has a load of benefits, but also has the trade off that all those modules are going to be loaded in the browser too, which makes things like `express` impossible to call.

In our squad, we use `React.lazy` to lazy load plugins and components and keep the initial js super lightweight, and then include the libraries built from source rather than built from `.cjs` which means you get treeshaking when it comes to packages being used on the client and used on the server.

I'm not saying it's a better way, but it might help mitigate having two seperate plugin packages.

I like the concept of plugins bringing their own api, which is basically just an express router which can be given it's own prefix in the `api` part.

```ts
export const router: Router = express.Router();

router.get('/v1/make-rest-great-again', (req: Request, res: Response) => res.send({done: true}));

// @backstage/backend
import {router as greatAgain} from '@backstage/plugin-great-again';

app.use('/great-again', greatAgain);
```

Reply:
but, totally agree with @Rugvip, let's keep with this approach for now as the talks around how to deal with this is something that shouldn't block work right now.

Reply:
Other reasons for not compiling to `.cjs` would be that the whole `hmr` and `react-fast-refresh` development experience becomes so much better. They can do static analysis to work out what needs to be recompiled rather than building the entire plugin again.